tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -554,6 +554,93 @@ int main(int argc, char **argv)
+
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_NOT_NULL(server_conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls13_client_config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, tls12_server_config));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair = { 0 };
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Negotiate initial handshake to produce TLS1.2 session ticket */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
+        EXPECT_TRUE(IS_ISSUING_NEW_SESSION_TICKET(server_conn));
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS12);
+
+        /* Store the TLS1.2 session ticket */
+        size_t tls12_session_ticket_len = s2n_connection_get_session_length(client_conn);
+        EXPECT_SUCCESS(s2n_alloc(&ticket, tls12_session_ticket_len));
+        EXPECT_SUCCESS(s2n_connection_get_session(client_conn, ticket.data, ticket.size));
+
+        /* Prepare client and server for a second connection */
+        EXPECT_OK(s2n_wipe_connections(client_conn, server_conn, &io_pair));
+
+        /* Client sets up a resumption connection with the received TLS1.2 session ticket data */
+        EXPECT_SUCCESS(s2n_connection_set_session(client_conn, ticket.data, ticket.size));
+
+        /* Set server config to TLS1.3 cipher preferences to allow TLS 1.3 */
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        /* Negotiate second connection */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS13);
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS13);
+
+        /* since TLS 1.3 was negotiated, and the client hasn't called recv yet
+         * there should be no session ticket available. 512 is an arbitrary
+         * number, nothing should be copied because this operation should fail
+         */
+        EXPECT_EQUAL(s2n_connection_get_session_length(client_conn), 0);
+        EXPECT_EQUAL(s2n_connection_get_session(client_conn, ticket.data, 512), 0);
+
+        /* the server automatically sends a session ticket as part of completing
+         * handshake io, so it isn't necessary to call send again
+         */
+        uint8_t data = 1;
+        s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+        EXPECT_BLOCKED_ON_IO(s2n_recv(client_conn, &data, 1, &blocked));
```

Comment:
I would expect you would need to NOT properly receive a new TLS1.3 ticket in order to repro the issue. Wouldn't  receiving a TLS1.3 ticket override client_ticket and also set the ticket_fields, resolving the problem?

But I could be wrong, and that'd probably make the issue more common. Have you tested your repro since you added this receive?

Reply:
If you define the bug as hitting the issue in set_session
```
s2n_connection_set_session() failed (SSL: invalid argument provided into a function call. ... 
```
then you're correct. The only failure in this test case are the asserts on lines 611 and 612. If a client calls recv before calling get_session, then they would not be impacted by this bug.

Reply:
Now that I'm typing that out, perhaps the third handshake is overzealous? I don't know that it's actually testing anything.

Reply:
Eh I'm torn. Like, the third handshake just seems irrelevant / unrelated. It would have succeeded with the bug in place.

It kind of seems like a different test. Like, you want to make sure that the server can't sabotage the ticket in some way, rather than the client. The server has arguably minimal contributions to the client's session ticket state (just the ticket_age_add), so it could mainly affect things by writing a bad psk identity or ticket_age_add, which does only get testing by doing another handshake.

Reply:
I left it in for now, since it seems reasonable to err on the side of more/deeper asserts. But let me know if you have a stronger opinion.

Reply:
More/deeper assert isn't always helpful though, or we'd have one giant test instead of separate test cases.

I went back and checked what you named the test case: "Test: s2n_connection_get_session should not return a ticket before it has been received if TLS 1.3 was negotiated." Based on that name, you definitely don't want to do another handshake. And if you do the extra handshake, I'm not sure what you would name the test case :)